### PR TITLE
modified the dockerfile for better caching

### PR DESCRIPTION
**Pull request by ultrons**
Split the copy operation into minimal setup and requirement.txt file and defer the full repo copy at the end.
This prevent full dependency rebuild every time a user makes local changes and wants to quickly build an re-run on a remote cluster.

### CODE REVIEW

**Review by gobbleturk — APPROVED** [review]
Thanks Vaibhav!

**Comment by jonb377** [comment]
cc @tonyjohnchen, I believe this is the use case of `docker_upload_runner.sh`

**Review by tonyjohnchen — APPROVED** [review]
LGTM

**Review by jonb377 — APPROVED** [review]

**Comment by ultrons** [comment]
Added one more fix. 
Jax jax lib installation should be done last in the setup.sh
Since we allow user to control it via the build args. 
The current order, install dependencies from requirements.txt last.
If any library in this list causes jax/jax lib version overwrite, the input from the the users will not be remain in the final build.
This leads to painful jax version debug.

To repro, you can build with JAX_VERSION=0.4.30 and MODE=stable
The final build will have jax/jaxlib and 0.4.33, which is not expected.
The reason is that jax version gets overwritten during the dependency installation via requirements.txt.

**Review by tonyjohnchen — APPROVED** [review]
lgtm

File: setup.sh [hunk 1]
```diff
@@ -86,6 +86,14 @@ fi
 # Save the script folder path of maxtext
 run_name_folder_path=$(pwd)
 
+# Install dependencies from requirements.txt
+cd $run_name_folder_path && pip install --upgrade pip
+if [[ "$MODE" == "pinned" ]]; then
+    pip3 install -U -r requirements.txt -c constraints_gpu.txt
+else
+    pip3 install -U -r requirements.txt
+fi
```
Comment (jonb377) [inline setup.sh:95]: Great catch!

**Review by jonb377 — APPROVED** [review]

**Pull request merged**